### PR TITLE
GEODE-6532: Use implementation, not compile, as build configuration

### PR DESCRIPTION
**Pull request by rhoughton-pivot**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [n/a] Have you written or updated unit tests to verify your changes?

- [n/a] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by rhoughton-pivot** [comment]
@PurelyApplied I have walked through the expected-pom.xml changes and looked at what all has moved from `compile` to `runtime` and nothing jumped out as a "But Robert, we use that symbol in a public API" kind of problem. The JIRA does call this out as a potential problem for downstream customers.

**Comment by rhoughton-pivot** [comment]
Just tested against `geode-examples` and saw that `commons-lang3` is no longer being emitted by us, and was expected by the example. Going back to inspect the packages to see if `geode` emits any symbols owned by `commons`

**Comment by Bill** [comment]
@rhoughton-pivot I pulled this branch and built it in IntelliJ and ran some JUnit tests in IntelliJ and didn't see any hiccups ✓

making some comments inline…

File: geode-gfsh/build.gradle [hunk 1]
```diff
@@ -21,26 +21,33 @@ apply from: "${project.projectDir}/../gradle/publish-java.gradle"
...
     api(platform(project(':boms:geode-all-bom')))
-    compile(project(':geode-core'))
-    compile(project(':geode-logging'))
-    compile(project(':geode-membership'))
-    compile(project(':geode-serialization'))
-    compile(project(':geode-unsafe'))
 
-    compile('com.healthmarketscience.rmiio:rmiio')
+    api(project(':geode-core'))
+    api(project(':geode-common'))
```
Comment (Bill) [inline geode-gfsh/build.gradle]: why `api` and not merely `implementation` here?

Comment (rhoughton-pivot) [reply geode-gfsh/build.gradle]: In geode-gfsh, the non-internal, public abstract class `org.apache.geode.management.cli.GfshCommand` has public methods such as `authorize` that takes a `ResourePermission.Resource`, defined in the  `geode-core` project. My understanding is, this makes `geode-core` an API-level dependency of `geode-gfsh`.

Similar reasoning for `geode-common`, in that it uses `org.apache.geode.annotations.Immutable` in a non-internal abstract class.

Comment (Bill) [reply geode-gfsh/build.gradle]: Golly, I didn't know that `GfshCommand` was meant to be extended by users. I see it's in a non-internal package. Does that mean we really have to actually support users extending that thing? @jinmeiliao do you have an opinion? Seems to me like it was a mistake to make `GfshCommand` non-internal.

Comment (onichols-pivotal) [reply geode-gfsh/build.gradle]: for a long time there was a vision to make Geode more modular, with each module providing its own gfsh commands (the ability for users to also write their own gfsh commands is sort of a side-effect of this...I too am curious if any customers actually have done so).  If we wanted to change course and say that gfsh is a fixed, non-extensible set of commands, we could also get rid of fast-classpath-scanner and a lot of other baggage...

Comment (jinmeiliao) [reply geode-gfsh/build.gradle]: Yes, gfsh commands are meant to be extensible, you can write your own gfsh commands and have them loaded in the command executor. `GfshCommand` is public for that reason.

Comment (Bill) [reply geode-gfsh/build.gradle]: ok thanks. @rhoughton-pivot this code is fine.

**Pull request merged**